File: web/src/@/lib/company-name.ts
/**
 * Display formatting for ASX company names.
 *
 * Two shapes reach the UI:
 *  1. Backend-title-cased names — services/shorts/.../postgres.go
 *     `cleanCompanyName` lowercases the ASIC name then title-cases it, which
 *     destroys ticker acronyms: "BHP GROUP LIMITED" becomes "Bhp Group",
 *     "CSL LIMITED" becomes "Csl". (Root cause; fixing it there needs a
 *     backend deploy AND a re-sync of company-metadata.)
 *  2. Raw ASIC / metadata names that are still SHOUTED ("BHP GROUP LIMITED").
 *
 * `formatCompanyName` repairs both at the display layer:
 *  - a word equal to the stock code (case-insensitive) always renders
 *    uppercase — "Bhp Group" + BHP → "BHP Group";
 *  - an all-caps source is title-cased, but 1-3 letter all-caps acronyms in
 *    that source keep their case — "BHP GROUP LIMITED" + BHP → "BHP Group";
 *  - trailing entity/security suffixes (LIMITED, LTD, ...) are dropped;
 *  - a mixed-case name is otherwise left alone — this never re-cases prose.
 *
 * Pure and dependency-free, so it is safe in both server and client chains.
 */

// Entity/security suffixes that add nothing to a display heading. The
// ORDINARY/ORD/CDI descriptors come from the ASIC PRODUCT field (e.g.
// "4DMEDICAL LIMITED ORDINARY"), which reaches the UI wherever names are read
// off the shorts table rather than company-metadata.
const TRAILING_SUFFIX =
  /[\s,]+(LIMITED|LTD\.?|CORPORATION|CORP\.?|INCORPORATED|INC\.?|PLC|N\.?L\.?|ORDINARY|ORD|CDI(\s+1:1)?)$/i;

/**
 * The ASIC PRODUCT field is `<company name> <security type> [qualifiers…]`:
 * "FIDUCIAN GROUP LTD ORDINARY FULLY PAID", "LENDLEASE GROUP FPO/UNITS
 * STAPLED", "BLOCK INC CDI 1:1 NYSE", "FLETCHER BUILDING ORD FOR. EXEMPT NZX",
 * "GRAINCORP LIMITED A CLASS ORDINARY". Everything from the FIRST security-type
 * token onward is instrument metadata, never part of the company name — so cut
 * there rather than trying to enumerate every trailing qualifier and venue.
 * Measured over 819 live ASIC names this takes the residual-descriptor rate
 * from 10.5% to 0%.
 */
// The trailing guard is "not followed by a letter" rather than \b, so the
// space-less "CDI1:1FOREXEMPT NYSE" form matches too (\b fails between the
// "I" and the "1", both word characters).
const SECURITY_TYPE =
  /[\s,]+(ORDINARY|ORD|FPO|CDI|UNITS?|STAPLED|NOTES?|[A-Z]\s+CLASS)(?![A-Za-z])/i;

const TOKENS = /([^A-Za-z0-9]+)/;

// Lowercased inside a title-cased name (never as the first word).
const MINOR_WORDS = new Set(["of", "and", "the", "for", "in", "on", "de"]);

/**
 * Upper-case the first LETTER, not the first character — a token can start
 * with a digit ("4DMEDICAL"), and capitalising position 0 there would leave
 * the whole word lowercase ("4dmedical" instead of "4Dmedical").
 */
function titleCaseWord(word: string): string {
  const firstLetter = word.search(/[A-Za-z]/);
  if (firstLetter < 0) return word;
  return (
    word.slice(0, firstLetter) +
    word.charAt(firstLetter).toUpperCase() +
    word.slice(firstLetter + 1).toLowerCase()
  );
}

/**
 * Does this word look like a mangled acronym rather than a real word?
 * A word with a vowel anywhere after the first letter is pronounceable
 * ("Rio", "Nab"), so we leave it alone; "Bhp"/"Csl"/"Anz" are not.
 * This keeps the code-match rule from SHOUTING legitimate names like
 * "Rio Tinto" (code RIO).
 */
function isAcronymLike(word: string): boolean {
  return !/[aeiou]/i.test(word.slice(1));
}

export function formatCompanyName(
  name: string | undefined | null,
  stockCode?: string,
): string {
  const code = stockCode?.trim().toUpperCase();
  if (!name) return code ?? "";

  let cleaned = name.trim();
  // Cut the instrument metadata tail first, so the entity-suffix strip below
  // sees a real trailing suffix ("SALUDA MEDICAL, INC. CDI USPROHEXCLQIB" →
  // "SALUDA MEDICAL, INC." → "Saluda Medical"). Never cut to nothing — a name
  // that IS a security word (or starts with one) keeps its original form.
  const securityCut = cleaned.search(SECURITY_TYPE);
  if (securityCut > 0) {
    const head = cleaned.slice(0, securityCut).trim();
    if (head) cleaned = head;
  }
  // Suffixes can stack ("… LIMITED" after "… CORPORATION"), and a trailing
  // full stop ("AGL Energy Limited.") would otherwise block the anchored
  // match, so drop trailing punctuation on each pass.
  let prev = "";
  while (prev !== cleaned) {
    prev = cleaned;
    cleaned = cleaned
      .replace(/[\s.]+$/, "")
      // A trailing parenthetical is a disambiguator, not part of the name,
      // and it blocks the anchored suffix match below:
      // "Environmental Group Limited (The)" -> "Environmental Group".
      .replace(/\s*\([^()]*\)$/, "")
      .replace(TRAILING_SUFFIX, "")
      .trim();
  }
  if (!cleaned) cleaned = name.trim();

  // Re-case when the source carries NO case information of its own — either
  // SHOUTED ("BHP GROUP") or entirely lower-case ("4dmedical", which is what
  // the pre-fix backend title-caser stored for digit-leading names, since it
  // capitalised character 0 instead of the first letter). A genuinely
  // mixed-case name is assumed to carry intentional casing and is left alone.
  const isShouting = cleaned === cleaned.toUpperCase();
  const isWhispering = cleaned === cleaned.toLowerCase();
  const needsRecasing = isShouting || isWhispering;

  let wordIndex = -1;
  const parts = cleaned.split(TOKENS);
  return parts
    .map((token, i) => {
      if (!/[A-Za-z]/.test(token)) return token; // separators / numbers
      wordIndex += 1;
      // A lone letter straight after an apostrophe is a possessive/contraction
      // suffix, never an acronym: "DOMINO'S" → "Domino's", not "Domino'S".
      // Applied even to a mixed-case source, because English never capitalises
      // it — and "Domino'S" is exactly the shape already stored in
      // company-metadata by the pre-fix backend title-caser.
      if (
        wordIndex > 0 &&
        token.length === 1 &&
        /['’]$/.test(parts[i - 1] ?? "")
      ) {
        return token.toLowerCase();
      }
      if (
        code &&
        token.toUpperCase() === code &&
        (isShouting || isAcronymLike(token))
      ) {
        return code;
      }
      if (!needsRecasing) return token;
      if (MINOR_WORDS.has(token.toLowerCase())) {
        // Leading minor word is title-cased ("THE A2 MILK" → "The A2 Milk"),
        // not lowercased and not treated as a short acronym.
        return wordIndex === 0 ? titleCaseWord(token) : token.toLowerCase();
      }
      // Short all-caps acronyms in a shouted source stay uppercase. Not
      // applied to a lower-case source, where a 3-letter token is far more
      // likely a word than an acronym.
      if (isShouting && token.length <= 3) return token.toUpperCase();
      return titleCaseWord(token);
    })
    .join("");
}

File: services/market-data-sync/stocklist/company_name.go
package stocklist

import (
	"regexp"
	"strings"
)

// cleanCompanyName turns a raw ASIC/ASX product name into a display-ready
// company name. It mirrors web/src/@/lib/company-name.ts formatCompanyName so
// the stored/served name matches what the frontend would repair at display
// time: trailing security/entity suffixes are stripped (they can stack), a
// SHOUTED source is title-cased with short all-caps acronyms (BHP, CSL, AGL)
// kept uppercase and minor words lowercased, and a word equal to the stock
// code always renders uppercase. A mixed-case source keeps its casing.
//
// NOTE: services/shorts/internal/store/shorts/company_name.go is an identical
// copy (separate Go module, no shared package) — change both together.
// securityTypeRe marks where instrument metadata begins in the ASIC PRODUCT
// field, which is `<company name> <security type> [qualifiers...]`:
// "FIDUCIAN GROUP LTD ORDINARY FULLY PAID", "LENDLEASE GROUP FPO/UNITS
// STAPLED", "BLOCK INC CDI 1:1 NYSE", "GRAINCORP LIMITED A CLASS ORDINARY".
// Everything from the FIRST security-type token onward is instrument detail,
// never part of the company name. The trailing guard is "not followed by a
// letter" so the space-less "CDI1:1FOREXEMPT NYSE" form matches too.
// Measured over 819 live ASIC names: residual descriptors 10.5%% -> 0%%.
var trailingParenRe = regexp.MustCompile(`\s*\([^()]*\)$`)

var securityTypeRe = regexp.MustCompile(`(?i)[\s,]+(ORDINARY|ORD|FPO|CDI|UNITS?|STAPLED|NOTES?|[A-Z]\s+CLASS)([^A-Za-z]|$)`)

func cleanCompanyName(name, stockCode string) string {
	code := strings.ToUpper(strings.TrimSpace(stockCode))
	cleaned := strings.TrimSpace(name)
	if cleaned == "" {
		// Keep empty empty: the stocklist upsert relies on NULLIF(name, '')
		// to avoid clobbering an existing good name.
		return ""
	}

	// Cut the instrument-metadata tail first, so the entity-suffix strip below
	// sees a real trailing suffix ("SALUDA MEDICAL, INC. CDI USPROHEXCLQIB" ->
	// "SALUDA MEDICAL, INC." -> "Saluda Medical"). Never cut to nothing.
	if loc := securityTypeRe.FindStringIndex(cleaned); loc != nil && loc[0] > 0 {
		if head := strings.TrimSpace(cleaned[:loc[0]]); head != "" {
			cleaned = head
		}
	}

	// Corporate-entity suffixes add nothing to a display heading. They stack
	// ("… CORPORATION LIMITED"), so trim until nothing changes.
	suffixes := []string{
		"ORDINARY", "ORD", "CDI 1:1", "CDI",
		"LIMITED", "LTD", "CORPORATION", "CORP",
		"INCORPORATED", "INC", "PLC", "NL",
	}
	// A trailing full stop ("AGL Energy Limited.") would otherwise block the
	// suffix match, so drop trailing punctuation on each pass.
	for {
		trimmed := strings.TrimRight(cleaned, " .")
		// A trailing parenthetical is a disambiguator, not part of the name,
		// and it blocks the suffix match below:
		// "Environmental Group Limited (The)" -> "Environmental Group".
		trimmed = strings.TrimSpace(trailingParenRe.ReplaceAllString(trimmed, ""))
		upper := strings.ToUpper(trimmed)
		for _, s := range suffixes {
			if strings.HasSuffix(upper, " "+s) || strings.HasSuffix(upper, ","+s) {
				trimmed = strings.TrimRight(trimmed[:len(trimmed)-len(s)], " ,")
				break
			}
		}
		if trimmed == cleaned {
			break
		}
		cleaned = trimmed
	}
	if cleaned == "" {
		cleaned = strings.TrimSpace(name)
	}

	// Re-case when the source carries NO case information of its own -- either
	// SHOUTED ("BHP GROUP") or entirely lower-case ("4dmedical", what the
	// pre-fix title-caser stored for digit-leading names). A genuinely
	// mixed-case name is assumed to be intentional and is left alone.
	isShouting := cleaned == strings.ToUpper(cleaned)
	isWhispering := cleaned == strings.ToLower(cleaned)
	needsRecasing := isShouting || isWhispering

	var b strings.Builder
	wordIndex := -1
	parts := splitCompanyTokens(cleaned)
	for i, token := range parts {
		if !strings.ContainsFunc(token, isASCIILetter) {
			b.WriteString(token) // separators / numbers
			continue
		}
		wordIndex++
		upperToken := strings.ToUpper(token)
		// A lone letter straight after an apostrophe is a possessive or
		// contraction suffix, never an acronym: "DOMINO'S" -> "Domino's".
		// Applied even to a mixed-case source, because English never
		// capitalises it.
		if wordIndex > 0 && len(token) == 1 && i > 0 && endsWithApostrophe(parts[i-1]) {
			b.WriteString(strings.ToLower(token))
			continue
		}
		switch {
		case code != "" && upperToken == code && (isShouting || isAcronymLike(token)):
			b.WriteString(code)
		case !needsRecasing:
			b.WriteString(token)
		case isMinorWord(token):
			// Leading minor word is title-cased ("THE A2 MILK" -> "The A2
			// Milk"), not lowercased and not treated as a short acronym.
			if wordIndex == 0 {
				b.WriteString(titleCaseWord(token))
			} else {
				b.WriteString(strings.ToLower(token))
			}
		case isShouting && len(token) <= 3:
			// Short all-caps acronyms in a shouted source stay uppercase.
			b.WriteString(upperToken)
		default:
			b.WriteString(titleCaseWord(token))
		}
	}
	return b.String()
}

// endsWithApostrophe reports whether a separator run ends in an apostrophe
// (straight or typographic).
func endsWithApostrophe(sep string) bool {
	return strings.HasSuffix(sep, "'") || strings.HasSuffix(sep, "\u2019")
}

func isASCIILetter(r rune) bool {
	return (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z')
}

func isAlphaNum(r rune) bool {
	return isASCIILetter(r) || (r >= '0' && r <= '9')
}

// splitCompanyTokens splits into alternating alphanumeric words and separator
// runs, preserving both (mirrors the /([^A-Za-z0-9]+)/ split in the web copy).
func splitCompanyTokens(s string) []string {
	var tokens []string
	start := 0
	var inWord bool
	for i, r := range s {
		w := isAlphaNum(r)
		if i == 0 {
			inWord = w
			continue
		}
		if w != inWord {
			tokens = append(tokens, s[start:i])
			start = i
			inWord = w
		}
	}
	if start < len(s) {
		tokens = append(tokens, s[start:])
	}
	return tokens
}

// Lowercased inside a title-cased name (never as the first word).
func isMinorWord(w string) bool {
	switch strings.ToLower(w) {
	case "of", "and", "the", "for", "in", "on", "de":
		return true
	}
	return false
}

// isAcronymLike reports whether a word looks like a mangled acronym rather
// than a real word: no vowel after the first letter ("Bhp", "Csl") — keeps
// the code-match rule from SHOUTING legitimate names like "Rio Tinto".
func isAcronymLike(w string) bool {
	return !strings.ContainsAny(w[1:], "aeiouAEIOU")
}

// titleCaseWord upper-cases the first LETTER, not the first character — a
// token can start with a digit ("4DMEDICAL"), and capitalising position 0
// there would leave the whole word lowercase ("4dmedical").
func titleCaseWord(w string) string {
	i := strings.IndexFunc(w, isASCIILetter)
	if i < 0 {
		return w
	}
	return w[:i] + strings.ToUpper(w[i:i+1]) + strings.ToLower(w[i+1:])
}
